endless/eosprofile.c

```diff
@@ -0,0 +1,567 @@
+
+  if (!probe->is_running)
+    return;
+
+  probe->is_running = FALSE;
+
+  /* Ideally, we just want to update the sample we just created, which means
+   * picking the last slot in the samples array; in practice, this is what
+   * should happen most of the time, unless we end up recursing. In that case
+   * we end up with creating a sample while the outer sample is still in
+   * flight.
+   *
+   * If we just recorded the inner samples, we'd end up with a skewed capture,
+   * as those samples would inevitably be masking the timing of the outer
+   * samples.
+   *
+   * The easiest approach is to discard the inner samples until we reach the
+   * outermost live sample
+   */
+  int first_in_flight = probe->samples->len - 1;
+  for (int i = probe->samples->len - 1; i >= 0; i--)
+    {
+      const ProfileSample *sample = &g_array_index (probe->samples, ProfileSample, i);
+
+      if (sample->end_time > 0)
+        break;
+
+      first_in_flight = i;
+    }
+
+  ProfileSample *sample = &g_array_index (probe->samples, ProfileSample, first_in_flight);
+
+  sample->end_time = sample_time;
+
+  if (first_in_flight != probe->samples->len - 1)
+    {
+      int range = probe->samples->len - first_in_flight - 1;
+
+      g_array_remove_range (probe->samples, first_in_flight + 1, range);
+    }
+}
+
+void
+eos_profile_state_init (void)
+{
+  static gboolean profile_state_inited;
+
+  if (profile_state_inited)
+    return;
+
+  profile_state_inited = TRUE;
+
+  const char *str = getenv ("EOS_PROFILE");
+  if (str != NULL)
+    {
+      profile_state = g_new0 (ProfileState, 1);
+      profile_state->probes = g_hash_table_new (g_str_hash, g_str_equal);
+
+      int capture_prefix_len = strlen ("capture");
+
+      if (g_ascii_strncasecmp (str, "capture", capture_prefix_len) == 0)
+        {
+          profile_state->capture = TRUE;
+
+          const char *filename = str + capture_prefix_len;
+          if (*filename == ':')
+            {
+              filename += 1;
+
+              if (*filename != '\0')
+                profile_state->capture_file = g_strdup (filename);
+            }
+
+          if (profile_state->capture_file == NULL || *profile_state->capture_file == '\0')
+            {
+              g_autofree char *capture_dir = g_build_filename (g_get_user_cache_dir (),
+                                                               "com.endlessm.Sdk.Profile",
+                                                               NULL);
+
+              if (g_mkdir_with_parents (capture_dir, 0700) < 0)
+                capture_dir = g_get_current_dir ();
+
+              profile_state->capture_file = g_strdup_printf ("%s%s%s-%d.db",
+                                                             capture_dir,
+                                                             G_DIR_SEPARATOR_S,
+                                                             g_get_prgname (),
+                                                             (int) getpid ()); 
```

Comment:
Do PIDs start from 1 in each new flatpak sandbox? If that's the case, this might not be so helpful.

Reply:
We could detect if it's a sandbox, and use something else, but ideally the sandboxed version should probably pass a full path.